### PR TITLE
Object store/ azure

### PR DESCRIPTION
**Pull request by deepsidhu85**
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

Added support for AZURE object store

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

Related to #228 

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

~* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.~
~* [ ] Tests added (or description of how to test) for any new features.~
~* [ ] User documentation updated for UI or technical changes.~

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -344,6 +344,11 @@
 			<artifactId>jackson-databind</artifactId>
 			<version>${jackson.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply pom.xml:348]: The azure sdk requires a package which is a part of the core jackson (>=2.10) and is not available as a separate maven package

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java [hunk 2]
```diff
@@ -35,6 +35,16 @@
...
+import com.google.common.net.HttpHeaders;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.hateoas.Link;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.ModelMap;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.web.servlet.view.RedirectView;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java:47]: Looks like they were missed. I will remove these

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	public File getTemporaryFile(Path file) {
+		File fileToProcess = null;
+
+		// We set the blobClient "path" to which we want to upload our file to
+		blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(file));
+
+		try {
+			// Create a file that will be unique in the /tmp/ folder. We append the current date/time
+			// to the file name
+			String tmpDir = "/tmp/" + new Date().toString().replaceAll("\\W", "");
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:61]: This has been completely changed. If you take a look at the object_store/_analysis_submissions PR, I've made use of the existing folders

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	/**
+	 * Removes the leading "/" from the absolute path
+	 * returns the rest of the path.
+	 *
+	 * @param file
+	 * @return
+	 */
+	private String getAzureFileAbsolutePath(Path file) {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:180]: Azure and AWS use a virtual filesystem. For example `/opt/irida/data/sequence/1/filename.fastq` would be the object. The file paths stored in the db have a leading slash which if we attempt to locate in the object store it tries to find containerName//object instead of containerName/object resulting in the bucket object (s3) or the blob(azure) not being found.

 I will update the comment to make this clear

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+	 */
+	@Override
+	public String getFileExtension(List<? extends SequencingObject> sequencingObjects) throws IOException {
+		String selectedExtension = null;
+		for (SequencingObject object : sequencingObjects) {
+
+			for (SequenceFile file : object.getFiles()) {
+				String fileName = getFileName(file.getFile());
+
+				Optional<String> currentExtensionOpt = VALID_EXTENSIONS.stream()
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:221]: Sure I will get that updated

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java [hunk 6]
```diff
@@ -65,6 +66,7 @@ public RESTProjectAnalysisController(ProjectService projectService,
 	 *         {@link Project}.
 	 */
 	@RequestMapping(value = "/api/projects/{projectId}/analyses", method = RequestMethod.GET)
+	@ResponseBody
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java:69]: Since the Azure sdk required jackson >2.10 the tests for this as well as the AnalysisAjaxTableController tests broke. My guess is the response is not being serialized properly

[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java:69]: Ok.  So I think something is going funny that made this necessary.  Our REST config in https://github.com/phac-nml/irida/blob/development/src/main/java/ca/corefacility/bioinformatics/irida/config/web/IridaRestApiWebConfig.java should serialize any resource to JSON automatically without the `@ResponseBody` annotation.  When I'm going through the rest api with curl I'm noticing some controllers being pretty-printed and some that aren't.  Pretty print is set up in https://github.com/phac-nml/irida/blob/development/src/main/java/ca/corefacility/bioinformatics/irida/config/web/IridaRestApiWebConfig.java#L73 and I think should be universal through the REST API.  

So long story short, something funny is going on to need this.  I think we'll have to dig because there's something going wrong with configuration somehow.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:61]: I see the `getFile` method in that branch, but it looks like it might have a slightly different use?  I understand the change later in the chain, but can it get pulled down to this PR since it was found here?  Its too easy for temporary stuff like this to accidentally slip by when getting reviewed later.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:180]: Ok.  Is there maybe a different implementation of `Path` that should be used for these cloud types?  Or maybe the `Path` class isn't the appropriate thing to store for this anymore?  I think `Path` is something that is assumed to be a file system reference.  Here since it's no longer a file on a file system we need to keep converting it.  Maybe we should be storing these paths in another way.  Lets talk about it next week.

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java:69]: Sure I will take a look into it and see if I can figure out what's going on

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:61]: Sure I can pull it down into here. The use is the same just renamed the method in a later branch

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java]: Removed in 23c0b6d48cdf222b4c5674c280f79ed361766082

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Updated in 23c0b6d48cdf222b4c5674c280f79ed361766082

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+/**
+ * Component implementation of file utitlities for azure storage
+ */
+@Component
+public class IridaFileStorageAzureUtilityImpl implements IridaFileStorageUtility {
+	private static final Logger logger = LoggerFactory.getLogger(IridaFileStorageAzureUtilityImpl.class);
+
+	private BlobServiceClient blobServiceClient;
+	private BlobContainerClient containerClient ;
+	private BlobClient blobClient;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Removed global in 23c0b6d48cdf222b4c5674c280f79ed361766082

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Updated in fdd25e1d2dc4f0342951f50e259cf2ec0657dfb2

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java]: So now the tests are passing..totally weird. Updated in 23c0b6d48cdf222b4c5674c280f79ed361766082

[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectAnalysisController.java]: Updated with PR #769

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+			InputStream initialStream = blobClient.openInputStream();
+			File targetFile = new File(file.toAbsolutePath().toString());
+			org.apache.commons.io.FileUtils.copyInputStreamToFile(initialStream, targetFile);
+			initialStream.close();
+			fileToProcess = targetFile;
+		} catch (BlobStorageException e) {
+			logger.trace("Couldn't find file on azure [" + e + "]");
+		} catch (IOException e) {
+			logger.debug(e.getMessage());
+		}
```
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:66]: As we discussed I think it would be good to wrap these exceptions in a `StorageException`.  We're already using that in the local storage utility so it would be good to use the same thing.

Please add that to any of the places in this file where we're just wrapping the `BlobStorageException` and logging.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+		// We set the blobClient "path" to which file we want to get
+		BlobClient blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(file));
+
+		try {
+			InputStream initialStream = blobClient.openInputStream();
+			File targetFile = new File(file.toAbsolutePath().toString());
+			org.apache.commons.io.FileUtils.copyInputStreamToFile(initialStream, targetFile);
+			initialStream.close();
+			fileToProcess = targetFile;
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:61]: As discussed lets try switching this to use `Files.createTempFile` instead of downloading to this specified location.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Updated in 4b16c81a35cbfe72080cf3818046d989db28fcad

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 9 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Updated in bdee3fd5ecbe7c0e5096d3f5889a51af59eac4f2

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+			Path tempFile = tempDirectory.resolve(file.getFileName().toString());
+			InputStream initialStream = blobClient.openInputStream();
+			org.apache.commons.io.FileUtils.copyInputStreamToFile(initialStream, tempFile.toFile());
+			initialStream.close();
+			fileToProcess = tempFile.toFile();
+		} catch (BlobStorageException e) {
+			logger.error("Couldn't find file on azure [" + e + "]");
+			throw new StorageException("Unable to locate file on azure", e);
+		} catch (IOException e) {
+			logger.error(e.getMessage());
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:70]: Updated in 5988e06b794923529c3dd6889ff54b983970516a

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void cleanupLocalFiles(Path path) {
+		logger.trace("Cleaning up temporary file downloaded from azure [" + path.toString() + "]");
+
+		Path origPath = path;
+		if(Files.isRegularFile(path)) {
+			origPath = path;
+			path = path.getParent();
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:87]: I don't think we should walk up to delete a parent directory.  While this shouldn't be an issue with our temp directories, it has the possibility of being accidentally misused and deleting a required directory.  if we're deleting a directory, I think it should be explicitly called to be deleted.

Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Updated in 05feefc11ed7c0e21deb78ad450036145ace7e5f

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleAssemblyController.java [hunk 12]
```diff
@@ -182,8 +183,7 @@ public ModelMap addNewAssemblyToSample(@PathVariable Long sampleId, @RequestPart
 		} finally {
 			// clean up the temporary files.
 			logger.trace("Deleted temp files");
-			Files.deleteIfExists(target);
-			Files.deleteIfExists(temp);
+			IridaFiles.cleanupLocalTemporaryFiles(target, temp);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleAssemblyController.java:186]: As discussed lets revert these changes back to the `deleteIfExists` stuff for now.  We can refactor this in a later branch when needed.

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/samples/SamplesAjaxController.java [hunk 13]
```diff
@@ -188,22 +199,24 @@ private void createSequenceFileInSample(MultipartFile file, Sample sample) throw
...
+	 * @return {@link IridaTemporaryFile}
 	 * @throws IOException Exception thrown if there is an error handling the file.
 	 */
-	private SequenceFile createSequenceFile(MultipartFile file) throws IOException {
+	private IridaTemporaryFile createSequenceFile(MultipartFile file) throws IOException {
 		Path temp = Files.createTempDirectory(null);
 		Path target = temp.resolve(file.getOriginalFilename());
 		file.transferTo(target.toFile());
-		return new SequenceFile(target);
+		return new IridaTemporaryFile(target, temp);
```
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/samples/SamplesAjaxController.java:220]: Yeah I think we should revert this to returning the `SequenceFile` for now.  It's back to the same issue of dealing with those temp directories.  We can address that in the future.

File: src/main/java/ca/corefacility/bioinformatics/irida/util/IridaFiles.java [hunk 14]
```diff
@@ -65,4 +66,34 @@ public static String getFileExtension(List<? extends SequencingObject> files) th
...
+		iridaFileStorageUtility.cleanupDownloadedLocalTemporaryFiles(filePath, directoryPath);
+	}
+
+	/**
+	 * Cleans up temporary files.
+	 *
+	 * @param filePath The path to the file
+	 * @param directoryPath The path to the directory which has the file
+	 */
+	public static void cleanupLocalTemporaryFiles(Path filePath, Path directoryPath) {
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/util/IridaFiles.java:85]: If it makes sense can we just leave this method accessable form the `IridaFileStorageUtility` class and not in `IridaFiles`?  That way it's only responsible for dealing with those temporarily downloaded files.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java [hunk 15]
```diff
@@ -16,15 +21,25 @@
...
+	public IridaTemporaryFile getTemporaryFile(Path file);
+
+	/**
+	 * Delete temporary downloaded file and/or directory.
+	 *
+	 * @param filePath The {@link Path} to the file
+	 * @param directoryPath The {@link Path} to the directory which has the file
 	 */
-	public File getTemporaryFile(Path file);
+	public void cleanupDownloadedLocalTemporaryFiles(Path filePath, Path directoryPath);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java:42]: For this method please accept an `IridaTemporaryFile` instead of the 2 `Path`s.  That will help ensure we only use this in the appropriate cases.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java [hunk 16]
```diff
@@ -106,4 +121,26 @@
...
 	 */
 	public String getFileExtension(List<? extends SequencingObject> sequencingObjects) throws IOException;
+
+	/**
+	 * Delete local temporary file and/or directory.
+	 *
+	 * @param filePath The {@link Path} to the file
+	 * @param directoryPath The {@link Path} to the directory which has the file
+	 */
+	public static void cleanupLocalTemporaryFiles(Path filePath, Path directoryPath) {
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java:131]: After reverting back to `Files.deleteIfExists` in those places we discussed, you should be able to remove this `static` method.  I think it should likely go away.

**Review by tom114 — CHANGES_REQUESTED** [review]

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaConcatenatorTemporaryFile.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+import java.nio.file.Path;
+
+import ca.corefacility.bioinformatics.irida.model.sequenceFile.SequencingObject;
+
+/**
+ * Used as a response for encapsulating a temporary file and it's directory
+ * for a concatenator object
+ */
+
+public class IridaConcatenatorTemporaryFile {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaConcatenatorTemporaryFile.java]: Removed this in 1dbdb47af3eedeee57410916380b562e92a85c41 as we will deal with cleaning up local temporary files in a future branch if needed

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/samples/SamplesAjaxController.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/samples/SamplesAjaxController.java]: Reverted in 1dbdb47af3eedeee57410916380b562e92a85c41

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleAssemblyController.java (hunk 12 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleAssemblyController.java]: Reverted in 1dbdb47af3eedeee57410916380b562e92a85c41

File: src/main/java/ca/corefacility/bioinformatics/irida/util/IridaFiles.java (hunk 14 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/util/IridaFiles.java]: Removed this method in 1dbdb47af3eedeee57410916380b562e92a85c41 as it is not required. We have the `cleanupDownloadedTemporaryFiles` method in the irida files storage utility classes to clean up any files downloaded from an object store

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java (hunk 15 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java]: Updated in 1dbdb47af3eedeee57410916380b562e92a85c41

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java (hunk 16 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java]: Removed in 1dbdb47af3eedeee57410916380b562e92a85c41

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaTemporaryFile.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
+package ca.corefacility.bioinformatics.irida.ria.web.dto;
+
+import java.nio.file.Path;
+
+/**
+ * Used as a response for encapsulating a temporary file and it's directory
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaTemporaryFile.java:6]: Can we update this comment to say something about how this is a temporarily downloaded file from an online service and should be cleaned up after it's done being used?

File: src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/Fast5Object.java [hunk 19]
```diff
@@ -103,7 +105,10 @@ private Fast5Type setType(SequenceFile file) {
 		try {
 			String extension = FilenameUtils.getExtension(getFile().getFileName());
 
-			if (file.isGzipped()) {
+			// Checks if file is where it should be before it checks if it is gzipped
+			if (IridaFiles.fileExists(file.getFile()) && file.isGzipped()) {
+				type = Fast5Object.Fast5Type.ZIPPED;
+			} else if (extension.equals("gz")) {
 				type = Fast5Object.Fast5Type.ZIPPED;
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/Fast5Object.java:112]: Got 2 parts of this if resulting in the same thing `ZIPPED`.  Should the extension checking be added as an `||` to the top condition?

File: src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java [hunk 20]
```diff
@@ -95,44 +98,62 @@ private void processSingleFile(SequenceFile sequenceFile) throws FileProcessorEx
 				.executionManagerAnalysisId(EXECUTION_MANAGER_ANALYSIS_ID)
 				.description(messageSource.getMessage("fastqc.file.processor.analysis.description", new Object[] {FastQCApplication.VERSION},
 						LocaleContextHolder.getLocale()));
+
+		IridaTemporaryFile iridaTemporaryFile = iridaFileStorageUtility.getTemporaryFile(fileToProcess);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java:102]: This processor gets a bit complicated.  Can you add comments through this file please about what's going on.  So up here something saying "we're getting a local copy of the files", down below "we're creating a temp directory", "we're running the fastqc modules", etc.

File: src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java [hunk 21]
```diff
@@ -95,44 +98,62 @@ private void processSingleFile(SequenceFile sequenceFile) throws FileProcessorEx
...
+		} catch (IOException e) {
+			logger.error("Unable to create temporary directory ", e);
+			throw new StorageException("Unable to create temporary directory", e);
+		} finally {
+				iridaFileStorageUtility.cleanupDownloadedLocalTemporaryFiles(iridaTemporaryFile);
+				try {
+					// Delete the analysis-output* temp directory
+					Files.deleteIfExists(outputDirectory);
+				} catch (IOException e) {
+					logger.error("Unable to remove directory", e);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java:155]: Should this exception be wrapped and re-thrown as a `StorageException` or `FileProcessorException` rather than just logging?

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+	 * {@inheritDoc}
+	 */
+	@Override
+	public IridaTemporaryFile getTemporaryFile(Path file) {
+		// We set the blobClient "path" to which file we want to get
+		BlobClient blobClient = containerClient.getBlobClient(getAzureFileAbsolutePath(file));
+
+		try {
+			logger.trace("Getting file from azure [" + file.toString() + "]");
+			Path tempDirectory = Files.createTempDirectory(null);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java:58]: Can we add a prefix here instead of `null`?  That will make things easier to debug if something is going wrong leaving temp directories around.  Maybe something like `"azure-tmp-"`.  We could add similar for the S3 stuff in its PR.

**Review by tom114 — CHANGES_REQUESTED** [review]
A couple last cleanup things.  I think this is looking very good, almost ready to go!

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java (hunk 22 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Sure good idea! Added in 117eedc7f06e586f08eed0fca9d520de62ca3c6f

File: src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java (hunk 21 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java]: Updated in 117eedc7f06e586f08eed0fca9d520de62ca3c6f

File: src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java (hunk 20 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/processing/impl/FastqcFileProcessor.java:103]: Added in 117eedc7f06e586f08eed0fca9d520de62ca3c6f

File: src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/Fast5Object.java (hunk 19 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/Fast5Object.java:110]: Yup I can't believe I did that. Updated in 117eedc7f06e586f08eed0fca9d520de62ca3c6f

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaTemporaryFile.java (hunk 18 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaTemporaryFile.java]: Updated in 117eedc7f06e586f08eed0fca9d520de62ca3c6f

Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaTemporaryFile.java]: Just noticed the package of this file.  Definitely shouldn't be in `dto`.  Probably more in something like `ca/corefacility/bioinformatics/irida/repositories/filesystem`

Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/dto/IridaTemporaryFile.java]: My bad. Been putting all dtos in that directory. Updated in d15b2a07ea240feaf6f39b90428d985fca96efa8

**Review by tom114 — APPROVED** [review]
